src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers/TypeUpgradeAnalyzer.cs

```diff
@@ -0,0 +1,169 @@
+        /// <param name="context">The syntax node analysis context including the identifier node to analyze.</param>
+        /// <param name="mappings">Type mappings to use when upgrading types.</param>
+        /// <param name="simpleName">The simple name of the identifier being analyzed.</param>
+        private static void AnalyzeIdentifier(SyntaxNodeAnalysisContext context, IEnumerable<TypeMapping> mappings, string simpleName)
+        {
+            // If the identifier isn't one of the mapped identifiers, bail out
+            var mapping = mappings.FirstOrDefault(m => m.SimpleName?.Equals(simpleName, StringComparison.Ordinal) ?? false);
+            if (mapping is null)
+            {
+                return;
+            }
+
+            // If the identifier resolves to an actual symbol that isn't the old identifier, bail out
+            if (context.SemanticModel.GetSymbolInfo(context.Node).Symbol is INamedTypeSymbol symbol
+                && !symbol.ToDisplayString(NullableFlowState.NotNull).Equals(mapping.OldName, StringComparison.Ordinal))
+            {
+                return;
+            }
+
+            // If the identifier is part of a fully qualified name and the qualified name exactly matches the new full name,
+            // then bail out because the code is likely fine and the symbol is just unavailable because of missing references.
+            var fullyQualifiedNameNode = context.Node.GetQualifiedName();
+            if (fullyQualifiedNameNode.ToString().Equals(mapping.NewName, StringComparison.Ordinal))
+            {
+                return;
+            }
+
+            // Store the new identifier name that this identifier should be replaced with for use
+            // by the code fix provider.
+            var properties = ImmutableDictionary.Create<string, string?>().Add(NewIdentifierKey, mapping.NewName);
+
+            // Create and report the diagnostic. Note that the fully qualified name's location is used so
+            // that the code fix provider can directly replace the node without needing to consider its parents.
+            var diagnostic = CreateDiagnostic(fullyQualifiedNameNode.GetLocation(), properties, simpleName, mapping.NewName);
+            context.ReportDiagnostic(diagnostic);
+        }
+
+        /// <summary>
+        /// Load type mappings from additional files.
+        /// </summary>
+        /// <returns>Type mappings as defined in *.typemap files in the project's additional files.</returns>
+        private static IEnumerable<TypeMapping> LoadMappings(ImmutableArray<AdditionalText> additionalTexts)
+        {
+            foreach (var file in additionalTexts)
+            {
+                if (file.Path.EndsWith(TypeMapFileSuffix, StringComparison.OrdinalIgnoreCase))
+                {
+                    foreach (var line in file.GetText()?.Lines ?? Enumerable.Empty<TextLine>())
+                    {
+                        var components = line.ToString().Trim().Split('\t');
```

Comment:
What about checking if it starts with `#` so people can do comments?

Reply:
Good idea. I'll add that.